corehq/apps/accounting/tests/test_customer_invoicing.py

```diff
@@ -0,0 +1,389 @@
+        """
+        invoice_start, invoice_end = get_previous_month_date_range(self.subscription.date_start)
+        invoice_factory = CustomerAccountInvoiceFactory(
+            account=self.account,
+            date_start=invoice_start,
+            date_end=invoice_end
+        )
+        invoice_factory.create_invoice()
+        invoice_set = self.subscription.invoice_set.union(self.sub2.invoice_set.all(), self.sub3.invoice_set.all())
+        self.assertEqual(invoice_set.count(), 0)
+
+    def test_no_invoice_after_end(self):
+        """
+        No invoices should be generated for the months after the end date of the subscriptions.
+        """
+        invoice_date = utils.months_from_date(self.sub2.date_end, 2)
+        invoice_start, invoice_end = get_previous_month_date_range(invoice_date)
+        invoice_factory = CustomerAccountInvoiceFactory(
+            account=self.account,
+            date_start=invoice_start,
+            date_end=invoice_end
+        )
+        invoice_factory.create_invoice()
+        invoice_set = self.subscription.invoice_set.union(self.sub2.invoice_set.all(), self.sub3.invoice_set.all())
+        self.assertEqual(invoice_set.count(), 0)
+
+
+class TestProductLineItem(BaseCustomerInvoiceCase):
+    """
+        Tests that the Product line item is properly generated in an invoice.
+        Customer level Invoice do not prorate monthly costs
+    """
+
+    def setUp(self):
+        super(TestProductLineItem, self).setUp()
+        self.product_rate = self.subscription.plan_version.product_rate
+
+    def test_product_line_items(self):
+        invoice_date = utils.months_from_date(self.subscription.date_start,
+                                              random.randint(2, self.subscription_length))
+        invoice_start, invoice_end = get_previous_month_date_range(invoice_date)
+        invoice_factory = CustomerAccountInvoiceFactory(
+            account=self.account,
+            date_start=invoice_start,
+            date_end=invoice_end
+        )
+        invoice_factory.create_invoice()
+        invoice_set = self.subscription.invoice_set.union(self.sub2.invoice_set.all(), self.sub3.invoice_set.all())
+        self.assertEqual(invoice_set.count(), 1)
+
+        invoice = invoice_set.first()
+        product_line_items = invoice.lineitem_set.get_products()
+        self.assertEqual(product_line_items.count(), 2)
+        for line_item in product_line_items:
+            self.assertTrue(line_item.base_description == 'One month of CommCare Advanced Software Plan.' or
+                            line_item.base_description == 'One month of CommCare Standard Software Plan.')
+            self.assertTrue(line_item.base_cost == self.product_rate.monthly_fee or
+                            line_item.base_cost == self.advanced_plan.product_rate.monthly_fee)
+
+
+class TestUserLineItem(BaseCustomerInvoiceCase):
```

Comment:
There's a third test case that should be tested: if the number of included users is N and two domains have X and Y users, where X < N and Y < N but X + Y > N, then the invoice should include a positive total.

I think the way to represent this is to have one user line item for the entire invoice, rather than one user line item per domain.

Reply:
Do you mean we should only ever have one sms line item? I though the point of having more than one sms line item per invoice was to have one sms line item per plan. One sms line item per plan would allow us to charge a user when they have >N SMSs among all their, say, Pro plans. This would also keep us from charging them if they have >N SMSs total (among all domains and all plans), but did not go over on any single plan.

Reply:
Ah, I missed that these were different plans.  What happens if you have multiple subscriptions on the same plan?  Is there one line item per subscription or per plan_version?

Reply:
There is one line item per plan_version.

Reply:
👍